lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -557,6 +558,98 @@ static StringRef getOperandName(OpOperand &oper, const hw::SymbolCache &syms,
   return buff;
 }
 
+/// Helper to get the circt.globalRef attribute.
+static ArrayAttr getGlobalRefs(Operation *op) {
+  return op->getAttrOfType<ArrayAttr>("circt.globalRef");
+}
+
+/// Helper to update GlobalRefOps after referenced ops bubble up.
+void PartitionPass::bubbleUpGlobalRefs(Operation *op) {
+  auto globalRefs = getGlobalRefs(op);
+  if (!globalRefs)
+    return;
+
+  for (auto globalRef : globalRefs.getAsRange<hw::GlobalRefAttr>()) {
+    // Resolve the GlobalRefOp and get its path.
+    auto refSymbol = globalRef.getGlblSym();
+    auto globalRefOp = dyn_cast_or_null<hw::GlobalRefOp>(
+        topLevelSyms.getDefinition(refSymbol));
+    assert(globalRefOp && "symbol must reference a GlobalRefOp");
+    auto oldPath = globalRefOp.namepath().getAsRange<hw::InnerRefAttr>();
+
+    // Construct a new path starting from the old path.
+    SmallVector<Attribute> newPath(oldPath);
+
+    // If there aren't two elements in the path, this GlobalRefOp has already
+    // been updated.
+    if (newPath.size() < 2)
+      return;
+
+    // Splice the last two elements in the path, using the combined name from
+    // the op's attributes, and taking the module from the second to last
+    // element.
+    newPath.pop_back_val().cast<hw::InnerRefAttr>();
+    auto nextElement = newPath.pop_back_val().cast<hw::InnerRefAttr>();
+    auto newModule = nextElement.getModule();
+    auto newNameAttr = StringAttr::get(op->getContext(), ::getOpName(op));
+    auto newLeaf = hw::InnerRefAttr::get(newModule, newNameAttr);
+    newPath.push_back(newLeaf);
+
+    // GlobalRefs use the inner_sym attribute, so keep it up to date.
+    op->setAttr("inner_sym", newNameAttr);
+
+    // Update the path on the GlobalRefOp.
+    auto newPathAttr = ArrayAttr::get(op->getContext(), newPath);
+    globalRefOp.namepathAttr(newPathAttr);
+  }
+}
+
+/// Helper to update GlobalRefops after referenced ops are pushed down.
+static void pushDownGlobalRefs(Operation *op, DesignPartitionOp partOp,
+                               DenseSet<Attribute> &newGlobalRefs) {
+  auto globalRefs = getGlobalRefs(op);
+  if (!globalRefs)
+    return;
+
+  for (auto globalRef : globalRefs.getAsRange<hw::GlobalRefAttr>()) {
+    // Resolve the GlobalRefOp and get its path.
+    auto rootModule = op->getParentOfType<ModuleOp>();
+    auto refSymbol = globalRef.getGlblSym();
+    auto globalRefOp = rootModule.lookupSymbol<hw::GlobalRefOp>(refSymbol);
```

Comment:
I think `lookupSymbol` does a linear search rather than looking up in a cache. Am I correct?

Reply:
Yep forgot to use the cache here.